tools/meteor-services/deploy.js

```diff
@@ -799,6 +814,88 @@ var listSites = function () {
   return 0;
 };
 
+// Given a hostname, add "http://" or "https://" as
+// appropriate. (localhost gets http; anything else is always https.)
+function addScheme(hostOrURL) {
+  if (hostOrURL.match(/^http/)) {
+    return hostOrURL;
+  } else if (hostOrURL.match(/^localhost(:\d+)?$/)) {
+    return "http://" + hostOrURL;
+  } else {
+    return "https://" + hostOrURL;
+  }
+};
+
+// Maps from "site" to deploy URL, so we don't have to re-ping on each RPC.
+const galaxyDiscoveryCache = new Map;
+
+// getDeployURL returns the base deploy URL for the given app.  "app" may be
+// falsey for certain RPCs (eg meteor list-sites).
+function getDeployURL(site) {
+  // Always trust explicitly configuration via env.
+  if (process.env.DEPLOY_HOSTNAME) {
+    return addScheme(process.env.DEPLOY_HOSTNAME);
+  }
+
+  const defaultURL = "https://galaxy.meteor.com";
+
+  // No site? Just use the default.
+  if (!site) {
+    return defaultURL;
+  }
+
+  // If we have a site, we can try to do Galaxy discovery.
+
+  // Do we already have an answer?
+  if (galaxyDiscoveryCache.has(site)) {
+    return galaxyDiscoveryCache.get(site);
+  }
+
+  // Try https first.  Production users should enable https so we will never
+  // fall back to http for them.
+  const urlFromHTTPS = discoverGalaxy(site, "https");
```

Comment:
Caching results of asynchronous operations raises a yellow flag, because if `getDeployURL` is called from another fiber while `discoverGalaxy` is suspended on the HTTP request, the second attempt would not benefit from the cache. The safe way to prevent such races is to use a `Promise<deploy URL>` as the value type of the `galaxyDiscoveryCache`, so that we can populate the cache with the promise immediately, and resolve the promise when the request comes back. In this case, however, I think it's safe to assume `getDeployURL` will never be called multiple times simultaneously, and extra HTTP requests aren't the end of the world anyway.